### ISSUE TITLE
Unable to obey repeat flags

### ISSUE THREAD

**Issue by flossy83** [HomeOfAviSynthPlusEvolution/L-SMASH-Works#24]
Test clip: https://drive.google.com/file/d/1lHB_EPFEKAQEiKab5LlYq0EyY9OOObeT/view?usp=share_link

Avisynth script:  `AudioDub(LWlibavVideoSource(file, repeat=true), LWLibavAudioSource(file, stream_index=-1))`

Error message: "LWLibavVideoSource: repeat requested for 59764336 frames by input video, but unable to obey (try repeat=false to get a VFR clip"

FFmpegSource2 is able to play the clip properly and obeys the repeat flags, however FFmpegSource2 doesn't update the frame property _FieldBased like LWLibav does, which is needed for proper deinterlacing (see 00:14 - 00:18 in the clip where the field order changes dynamically).

Relevant forum post: https://forum.doom9.org/showpost.php?p=1978668&postcount=1427

Thank you

**Comment by AkarinVS** [comment]
ffms2 couldn't possibly obey the repeat request as it doesn't have the feature at all.
if soft pulldown is used in the source, then it will return a vfr clip. If you want the ffms2 behavior, just pass `repeat=false`.

**Comment by AkarinVS** [comment]
That being said, my VS port doesn't have issues obeying the repeat flag in the example file you provided. I think the avs portion must have some bug (uninitialized variables? the number of frames is clearly bogus).

**Comment by kedaitinh12** [comment]
Yes we need Asd-g fix it 😂😂😂

**Comment by Asd-g** [comment]
@AkarinVS, with version `vA.3j` and `lsmas.LWLibavSource(source=r"480i multiple cadences.mkv", repeat=1)` I have the same error `Python exception: lsmas: repeat requested for 2097 frames by input video, but unable to obey (try repeat=0 to get a VFR clip).`

**Comment by flossy83** [comment]
Here is another clip showing the same error when using repeat=true.

https://rationalqm.us/misc/lainvob.vob

Unfortunately repeat=true is needed otherwise the video playback speed is wrong.   But that behaviour is wrong too as DirectShowSource() doesn't obey repeat fields and that plays it correctly (DirectShow is configured to use LAV Splitter, LAV Audio Decoder, LAV Video Decoder on my system)

**Comment by flossy83** [comment]
Oh I forgot we can put (fpsnum=30000, fpsden=1001) and then (repeat=true) works

**Comment by Asd-g** [comment]
Did you read [README and the description of the `field` parameter](https://github.com/HomeOfAviSynthPlusEvolution/L-SMASH-Works/blob/master/AviSynth/README#L267-L271)?

LWLibavVideoSource("lainvob.vob") - you have 1762 frames and 29.970 fps. Default repeat is unspecified - `silently returning a VFR clip with a constant (but wrong) fps`.

When you specify `fpsnum=30000, fpsden=1001`, this is VFR->CFR conversion and `Note that this option is ignored when VFR->CFR conversion is enabled.` (from README)

**Comment by flossy83** [comment]
Thanks, and sorry I didn't know that as I was reading outdated documentation at http://avisynth.nl/index.php/LSMASHSource/LWLibavVideoSource

**Comment by Asd-g** [comment]
Btw why don't you use [D2VSource](https://github.com/Asd-g/MPEG2DecPlus) (read README.md for info about the frame properties) or [DGDecode](http://avisynth.nl/index.php/DGDecode) for these MPEG-2 videos?

[Here](https://github.com/Asd-g/MPEG2DecPlus/issues/9) the discussion about the frame properties.

**Comment by flossy83** [comment]
Thanks... but what is the advantage?   Because LWlibav already updates _FieldBased and GetParity per-frame.   The only thing stopping me from using it on certain files is the error message bug.

Also, I really don't want to spend the week indexing my DVD collection to d2v :)

**Comment by flossy83** [comment]
Sorry, I was confused again.  OP bug only occurs with repeat=true, and I was talking about scenario where repeat=false.   

And, I've found some episodes where _FieldBased=progressive on frame with orphaned field, so that won't be reliable either.   So I will still have to use manual combing detection.

Despite this, it would still be nice to have the bug mentioned in OP fixed just in case it occurs with some other videos.  But, it seems to only affect short clips.

**Comment by flossy83** [comment]
> When you specify `fpsnum=30000, fpsden=1001`, this is VFR->CFR conversion and `Note that this option is ignored when VFR->CFR conversion is enabled.` (from README)

Just to note... when I am using `LWlibavVideoSource(repeat=false, fpsnum=24000, fpsden=1001)`  it doesn't perform the same as `FFMS2(rffmode=2)`.    The output from FFMS2 is smooth;  LWlibav is intermittently stuttery.    Renderer reports 23.976fps for both.

If I don't specify fpsnum & fpsdenum, LWlibav outputs 24.176 which causes the audio to go out of sync after a minute or two.  For debugging purposes the short clips are not reliable, so I've uploaded a longer clip [here](https://drive.google.com/file/d/1xz_WEX36iv8tpC18zR8e8-etGwtqfuVb/view?usp=share_link) (remux from my physical copy of the disc which was released in [June 2022](https://viavision.com.au/shop/caroline-in-the-city-the-complete-series/)).

For comparison, without Avisynth (just using MPC-HC with LAV Video Decoder and MadVR renderer) repeat flags are also not obeyed, and output is smooth with no sync issue. [According to LAV author nevcairiel](https://forum.doom9.org/showthread.php?p=1977836#post1977836 ): 
> LAV does not repeat fields on its own. It forwards the metadata to the renderer, which can decide to repeat frames, or not to. 

So in that case it seems the renderer is actually adjusting the timing.

So to summarise my observation for soft telecine content where I do not wish to obey repeat fields (for the purpose of preserving progressive frames to avoid having to redundantly IVTC them):

LAV Video Decoder + MadVR = smooth
FFMS2(rffmode=2) + MadVR = smooth
LwLibavVideoSrouce(repeat=false, fpsnum=24000, fpsden=1001) + MadVR = intermittent stutter
DirectShowSource(convertFPS=true, fps=23.976023) + MadVR = intermittent stutter

**Comment by Asd-g** [comment]
`LWlibavVideoSource(repeat=false, fpsnum=24000, fpsden=1001)` is equivalent of `FFVideoSource(rffmode=0, fpsnum=24000, fpsden=1001)`.

`FFMS2(rffmode=2)` is something like `LWlibavVideoSource(repeat=false)/FFVideoSource(rffmode=0)` + `TimecodeFPS("timecodes", 24000, 1001, start=true)`.

You could do:
1) `FFVideoSource(timecodes="timecodes.txt")` and run analyze pass to create the timecodes file.
2) Then open the video `LWlibavVideoSource(repeat=false)` - vfr.
3) Then do the filtering.
4) As a final step do VFR->CFR `TimecodeFPS("timecodes", 24000, 1001, start=true)`.

**Comment by flossy83** [comment]
Interesting, thank you for the tip.

I tried also just playing the clip with FFMS2(repeat=2) and using the LWlibav clip just for getting frame props to apply to FFMS2 clip, but the sync between the two clips was not always perfect.

**Comment by flossy83** [comment]
Sorry to bump this, but is there any chance in future of possibly getting FFMS2's rffmode=2 (DVD2AVI's "force film" mode) implemented into LWLibavVideoSource?     

This would be a major boon for my growing collection of NTSC DVD's which use soft telecine.  For now I'm having to use FFMS2, but would prefer LWLibav.

**Comment by Asd-g** [comment]
`LWlibavVideoSource(repeat=false) + TimecodeFPS` isn't ok for you?

**Comment by flossy83** [comment]
It's only an issue of practicality for me.   But I think my assumption that DVD2AVI is a good solution, may be wrong?   I don't even really know what is the "correct" way to play soft telecine in such a way that we ignore repeat flags and get progressive output that is still in sync with audio.    LAV Video Decoder + MadVR gets it right, but I have no idea how it works.    I don't think it is the same as DVD2AVI.
